### PR TITLE
Mat 4203 storing umls key

### PR DESCRIPTION
**Pull request by sb-cecilialiu**
## Terminology Service PR

Jira Ticket: [MAT-4203](https://jira.cms.gov/browse/MAT-4203)
(Optional) Related Tickets:

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [x] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

**Comment by sb-cecilialiu** [comment]
This branch is for saving UMLS user info so front end doesn't need to store and pass in TGT to backend. The branch does not include encryption, so it won't be merged.

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java [hunk 1]
```diff
@@ -32,23 +41,28 @@ public class VsacController {
 
   @GetMapping(path = "/valueSet", produces = MediaType.APPLICATION_JSON_VALUE)
   public ResponseEntity<String> getValueSet(
-      @RequestParam(name = "tgt") String tgt,
-      @RequestParam(name = "oid") String oid,
+      Principal principal,
+      @RequestParam(required = true, name = "oid") String oid,
```
Comment (RohitKandimalla) [inline src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: I don't think we have to specify the required attribute to "true". by default, all method parameters annotated with @RequestParam are required by default.

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java [hunk 2]
```diff
@@ -59,7 +73,42 @@ protected String serializeFhirValueset(ValueSet fhirValueSet) {
...
+    Optional<UmlsUser> umlsUser = vsacService.findByHarpId(username);
+    if (umlsUser.isPresent() && umlsUser.get().getApiKey() != null) {
+      return ResponseEntity.ok().body(vsacService.validateCodes(cqlCodes, umlsUser.get().getTgt()));
+    }
+    return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+  }
+
+  @PostMapping(path = "/umlsLogin")
+  public ResponseEntity<String> umlsLogin(
+      Principal principal, @RequestParam(required = true, name = "apiKey") String apiKey)
```
Comment (RohitKandimalla) [inline src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: The same here "required=true" is redundant

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java [hunk 3]
```diff
@@ -59,7 +73,42 @@ protected String serializeFhirValueset(ValueSet fhirValueSet) {
...
   public ResponseEntity<List<CqlCode>> validateCodes(
-      @RequestBody List<CqlCode> cqlCodes, @RequestParam String tgt) {
-    return ResponseEntity.ok().body(vsacService.validateCodes(cqlCodes, tgt));
+      Principal principal, @RequestBody List<CqlCode> cqlCodes) {
+    final String username = principal.getName();
+    Optional<UmlsUser> umlsUser = vsacService.findByHarpId(username);
+    if (umlsUser.isPresent() && umlsUser.get().getApiKey() != null) {
+      return ResponseEntity.ok().body(vsacService.validateCodes(cqlCodes, umlsUser.get().getTgt()));
+    }
+    return new ResponseEntity<>(HttpStatus.NOT_FOUND);
```
Comment (RohitKandimalla) [inline src/main/java/cms/gov/madie/terminology/controller/VsacController.java:93]: Do you think it's useful to return 401 instead of 404?

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java [hunk 4]
```diff
@@ -59,7 +73,42 @@ protected String serializeFhirValueset(ValueSet fhirValueSet) {
...
+    log.debug("Entering: umlsLogin(): username = " + username);
+
+    String tgt = vsacService.getTgt(apiKey);
+    UmlsUser umlsUser = vsacService.saveUmlsUser(username, apiKey, tgt);
+    String msg = "User: " + umlsUser.getHarpId() + " successfully loggin in to UMLS.";
+    log.debug("msg = ");
+    return ResponseEntity.ok().body(msg);
+  }
+
+  @GetMapping("/checkLogin")
```
Comment (RohitKandimalla) [inline src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: It is preferred to use lowercase letters in URI paths, I think we can do check-login or check/login

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 2 above)
Comment (nmorasb) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: Could we form a URI with lowercase characters, also avoiding verbs and trying to form using nouns for resources so something like `/umls-credentials`

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 4 above)
Comment (nmorasb) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: agreed, a URI with lowercase characters, also avoiding verbs and trying to form using nouns for resources so something like `GET /umls-credentials/status`

Comment (sb-cecilialiu) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: @RohitKandimalla Thanks a lot Rohit! I'm changing uri to lower cases.

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 3 above)
Comment (sb-cecilialiu) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java:93]: @RohitKandimalla For this one, as UMLS returns 401 when getting a service ticket, so I used 404 just to differentiate.

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 2 above)
Comment (sb-cecilialiu) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: @nmorasb Thanks Nate! So for login to umls, we should use /umls-credentials?

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 3 above)
Comment (RohitKandimalla) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java:93]: Getting an API-KEY or TGT or service ticket, all 3 comes under authentication. 404 indicates that "code or code system is not found" but that's not the case.

File: pom.xml [hunk 5]
```diff
@@ -87,6 +87,17 @@
...
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-data-mongodb</artifactId>
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/org.mongodb/mongodb-crypt -->
+    <dependency>
+      <groupId>org.mongodb</groupId>
+      <artifactId>mongodb-crypt</artifactId>
+      <version>1.4.0</version>
+    </dependency>
```
Comment (jkotanchik-SB) [inline pom.xml]: Can `mongodb-crypt` be removed for now?

Comment (sb-cecilialiu) [reply pom.xml]: @jkotanchik-SB Yeah it should be removed. I just did. Thanks!

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 3 above)
Comment (sb-cecilialiu) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java:93]: @RohitKandimalla This is to tell that there is no database record for the user. That's why.

Comment (jkotanchik-SB) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java:93]: >401 Unauthorized response status code indicates that the client request has not been completed because it lacks valid authentication credentials for the requested resource.

That aligns with a user attempting access a resource, in this case a Code on VSAC, and because they don't have an api key available in the system, they do not have the proper authn creds to access that resource.

A 404 here would indicate the target Code was not located in VSAC, and will likely be reported as such to the user.

This does point out an internal edge case. What if the user manages to request code validation without logging into umls (and thereby persisting their api key)?

Alternatively, maybe a 400 error with a message similar to "Unable to complete request, missing credentials". That would clarify for the caller the root cause of the failure.

Comment (jkotanchik-SB) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java:93]: There's no complimentary error in the value set validation method. Any idea what it returns when the user's creds are missing?

**Review by jkotanchik-SB — CHANGES_REQUESTED** [review]
Looks good, just need to come to a consensus on how to handle missing creds during validation/expansion requests.

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 3 above)
Comment (sb-cecilialiu) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java:93]: @jkotanchik-SB @RohitKandimalla I changed it to unauthorized, if user credentials are missing.

**Review by jkotanchik-SB — DISMISSED** [review]

File: src/main/java/cms/gov/madie/terminology/webclient/TerminologyServiceWebClient.java [hunk 6]
```diff
@@ -111,4 +122,28 @@ public VsacCode getCode(String codePath, String serviceTicket) {
             })
         .block();
   }
+
+  public String getTgt(String apiKey) throws InterruptedException, ExecutionException {
+    String uri = String.format(baseUrl + utsLoginEndpoint, apiKey);
+    Mono<String> responseMono =
+        terminologyClient
+            .post()
+            .uri(uri)
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/terminology/webclient/TerminologyServiceWebClient.java]: VSAC and UMLS recommends sending the api-key in the request body. It looks it is being sent in the URL. Is that correct?

Comment (RohitKandimalla) [reply src/main/java/cms/gov/madie/terminology/webclient/TerminologyServiceWebClient.java]: It could be my mistake, I never verified it by passing it as a body, the confluence doc specifies to send it as a param. but looks like we can send the API key either way. 

Yes we can send the api-key in the body 

image

Comment (sb-cecilialiu) [reply src/main/java/cms/gov/madie/terminology/webclient/TerminologyServiceWebClient.java]: @jkotanchik-SB @RohitKandimalla I changed it to send through request body.

**Review by RohitKandimalla — APPROVED** [review]

**Review by jkotanchik-SB — APPROVED** [review]

File: src/main/java/cms/gov/madie/terminology/controller/VsacController.java (hunk 2 above)
Comment (sb-cecilialiu) [reply src/main/java/cms/gov/madie/terminology/controller/VsacController.java]: Changed it to use umls-credentials
